cumulus_library/apis/umls.py

```diff
@@ -0,0 +1,152 @@
+    "snomed-ct-us-edition-transitive-closure-resources",
+    "snomed-ct-international-edition",
+    "snomed-ct-core-problem-list-subset",
+    "snomed-ct-to-icd-10-cm-mapping-resources",
+    "snomed-ct-spanish-edition",
+    "umls-metathesaurus-full-subset",
+    "umls-metathesaurus-mrconso-file",
+    "umls-full-release",
+]
+
+
+class UmlsApi:
+    def __init__(self, api_key: str | None = None, validator_key: str | None = None):
+        """Creates a requests session for future calls, and validates the API key
+
+        :keyword api_key: A UMLS API key (will check for an env var named UMLS_API_KEY
+            if None)
+        :keyword validator_key: A UMLS API key for the calling application. Will be
+            set to the value of api_key if None (which is the current expected
+            behavior, since we don't want to be distributing UMLS keys)
+        """
+
+        if api_key is None:
+            api_key = os.environ.get("UMLS_API_KEY", None)
+            if api_key is None:
+                raise errors.ApiError("No UMLS API key provided")
+        self.api_key = api_key
+        if validator_key is None:
+            validator_key = api_key
+        self.validator_key = validator_key
+
+        auth_payload = {"validatorApiKey": self.validator_key, "apiKey": self.api_key}
+        self.session = requests.Session()
+        response = self.session.get(
+            "https://utslogin.nlm.nih.gov/validateUser", params=auth_payload
+        )
+        if response.status_code == 401:
+            raise errors.ApiError("Invalid UMLS API validator key")
+        if response.text != "true":
+            raise errors.ApiError("Invalid UMLS API key")
+        self.session.auth = requests.auth.HTTPBasicAuth("apikey", api_key)
+
+    def get_vsac_valuesets(
+        self, url: str | None = None, oid: str | None = None
+    ) -> list[dict]:
+        """Gets a valueset, and any nested valuesets, from the VSAC API
+
+        :keyword url: an URL to target for a valueset (typically expected)
+        :keyword oid: A valuset OID
+        :returns: A list, containing the valueset and any referenced
+            valuesets.
+
+        Documentation on this API is available at
+        https://www.nlm.nih.gov/vsac/support/usingvsac/vsacfhirapi.html
+
+
+        TODO: do we need to support the FHIR operators?
+        TODO: do we need to support the v2 API?
+        https://www.nlm.nih.gov/vsac/support/usingvsac/vsacsvsapiv2.html
+        """
+        if url is None:
+            url = "https://cts.nlm.nih.gov/fhir/res/ValueSet"
+        if oid:
+            url = f"{url}/{oid}"
+
+        # If we're inspecting url references in a VSAC response, they come back
+        # specifying a url that does not align with the actual implemented rest
+        # APIs, so we do some massaging
+        if "http:" in url:
+            url = url.replace("http:", "https:")
+        if "/res/" not in url:
+            url = url.replace("/fhir/", "/fhir/res/")
```

Comment:
That's rude

Reply:
the data model is sacred and must never be modified